core/endorser/endorser.go

```diff
@@ -458,9 +464,56 @@ func (e *Endorser) ProcessProposalSuccessfullyOrError(up *UnpackedProposal) (*pb
 		Endorsement: endorsement,
 		Payload:     mPrpBytes,
 		Response:    res,
+		Interest:    ccInterest,
 	}, nil
 }
 
+// Using the simulation results, build the ChaincodeInterest structure that the client can pass to the discovery service
+// to get the correct endorsement policy for the chaincode(s) and any collections encountered.
+func (e *Endorser) buildChaincodeInterest(simResult *ledger.TxSimulationResults) (*pb.ChaincodeInterest, error) {
+	ccInterest := &pb.ChaincodeInterest{}
+
+	chaincodes := map[string]struct{}{}
+	if simResult.PvtSimulationResults != nil {
+		// list collections that are read from (in order to populate NoPublicReads in ChaincodeInterest)
+		for _, prws := range simResult.PvtSimulationResults.GetNsPvtRwset() {
+			for _, cprws := range prws.GetCollectionPvtRwset() {
+				// Since each collection in a chaincode could have different values of the NoPrivateReads flag, create a new Chaincode entry for each.
+				ccInterest.Chaincodes = append(ccInterest.Chaincodes, &pb.ChaincodeCall{
+					Name:            prws.GetNamespace(),
+					CollectionNames: []string{cprws.GetCollectionName()},
+					NoPrivateReads:  !simResult.PrivateReads.Exists(prws.Namespace, cprws.CollectionName),
+				})
+				chaincodes[prws.GetNamespace()] = struct{}{}
+			}
+		}
+	}
```

Comment:
I missed noticing in this block earlier
- Even in the private read-write, we should skip the _lifecycle namespace (can skip IsSysCC(nsrws.GetNamespace())) -- As the client would specifically target the orgs for these transactions.
- This logic assumes that if a collection is read from, it is written to as well. So, the collections that are only present in the `PrivateReads` and not in the pvt-write-set will get a miss.

Reply:
Good catch.  This logic was left over from my prototype which was building a full private read set, so this assumption would have held.  I've now rewritten this entirely in a (hopefully) clearer way.  And added an extra test case for this situation.